pkg/verify/tlog.go

```diff
@@ -35,14 +35,39 @@ import (
 
 const maxAllowedTlogEntries = 32
 
+type GetRekorClientFunc func(string) (*rekorGeneratedClient.Rekor, error)
+type verifytlogOptions struct {
+	getRekorClient GetRekorClientFunc
+}
+
+func makeOptions(opts ...TlogOption) verifytlogOptions {
+	opt := verifytlogOptions{}
+	for _, o := range opts {
+		o(&opt)
+	}
+	return opt
+}
+
+// TlogOption is a functional option for transparency log verification.
+type TlogOption func(*verifytlogOptions)
+
+// WithGetRekorClientFunc sets the function that will be used to fetch rekor client from base URL.
+// If not provided, rekorClient.GetRekorClient is used.
+func WithGetRekorClientFunc(f GetRekorClientFunc) TlogOption {
+	return func(opts *verifytlogOptions) {
+		opts.getRekorClient = f
+	}
+}
+
 // VerifyArtifactTransparencyLog verifies that the given entity has been logged
 // in the transparency log and that the log entry is valid.
 //
 // The threshold parameter is the number of unique transparency log entries
 // that must be verified.
 //
 // If online is true, the log entry is verified against the Rekor server.
-func VerifyArtifactTransparencyLog(entity SignedEntity, trustedMaterial root.TrustedMaterial, logThreshold int, trustIntegratedTime, online bool) ([]Timestamp, error) { //nolint:revive
+func VerifyArtifactTransparencyLog(entity SignedEntity, trustedMaterial root.TrustedMaterial, logThreshold int, trustIntegratedTime, online bool, opts ...TlogOption) ([]Timestamp, error) { //nolint:revive
```

Comment:
This doesn't feel like the optimal way to use the Options pattern to me, it adds a lot of lines of code just to allow the test to switch to a mock client and there are no other options. Could we instead add an argument to pass in the client directly, for instance like https://github.com/sigstore/cosign/blob/677a262c3205c7bf8612f30b7b44bdf51bd68bac/internal/pkg/cosign/tsa/signer.go#L38 ?

This library is pre-1.0 so it's okay to change the function signature if there is a good reason for it and if you mention it in the release notes.

Reply:
The reason why I did this is because: 
1. `VerifyArtifactTransparencyLog` generates a new rekor client for every entry in the entities tlog, therefore one rekor client might not work if base URLs are different. 
2. Having rekor client (or a func to create rekor client) as a required argument can be confusing for users of the library, as, in my opinion, they should not be expected to figure out what client to create

what do you think?